### ISSUE TITLE
prometheusreceiver: Unable to relabel job and instance labels using relabel_configs and metric_relabel_configs in prometheus receiver

### ISSUE THREAD

**Issue by rashmichandrashekar** [open-telemetry/opentelemetry-collector-contrib#5663]
**Describe the bug**
When using relabel_config to relabel job or using metric_relabel_configs to relabel job/instance labels, the collector fails to scrape metrics

**Steps to reproduce**
Provide target_label as job in the relabel_config or job/instance in metric_relabel_configs.

**What did you expect to see?**
Job/instance label should be replaced with the desired replacement value

**What did you see instead?**
When using relabel_config with target_label job, saw this error - 
{"level":"warn","ts":1633406450.405102,"caller":"scrape/scrape.go:1104", "msg":"Appending scrape report failed","kind":"receiver","name":"prometheus","scrape_pool":"prometheus_ref_app",  "target":"http://10.244.1.76:2112/metrics","err":"unable to find a target group with job=job_replacement"}

When using metric_relabel_config saw this error - 
{"level":"debug","ts":1633485477.928548,"caller":"scrape/scrape.go:1355","msg":"Unexpected error","kind":"receiver",
        # "name":"prometheus","scrape_pool":"prometheus_ref_app","target":"http://10.244.1.76:2112/metrics",
        # "series":"go_gc_duration_seconds{quantile=\"0\"}","err":"unable to find a target with job=prometheus_ref_app, 
        # and instance=instance_replacement"}

**What version did you use?**
Version: v0.27.0

**What config did you use?**
Config: (e.g. the yaml config file)

    global:
      evaluation_interval: 5s
      scrape_interval: 5s
    scrape_configs:
    scrape_configs:
    - job_name: prometheus_ref_app
      scheme: http
      kubernetes_sd_configs:
        - role: pod
      relabel_configs:
        - source_labels: [__meta_kubernetes_pod_label_app]
          action: keep
          regex: "prometheus-reference-app"
        - source_labels: [__address__]
          replacement: job_replacement
          target_label: job
     metric_relabel_configs:
        - source_labels: [__address__]
          replacement: instance_replacement
          target_label: instance

**Environment**
OS:"Ubuntu 20.04"
Compiler(if manually compiled): go 1.14

**Comment by rashmichandrashekar** [comment]
@dashpole - Sorry for the delay. I tried it with prometheus and I see that the metadata seems to be available when i tried a few targets. I queries the target metadata api to view the metadata. Was that what you were looking for?
For one of the tests I configured 2 scrape jobs for the metrics_path - /metrics/cadvisor - one with job relabelled and the other without and I see metadata for both - 
![image](https://user-images.githubusercontent.com/36861737/144343918-27859fdf-49a4-431a-8632-d28ea5f066ba.png)

**Comment by dashpole** [comment]
Thats really interesting.  In theory, we should be querying the target metadata when we go check for metadata.  It probably needs more digging

**Comment by dashpole** [comment]
# metric relabel configs + job

## With prometheus config:

```yaml
scrape_configs:
- job_name: 'metric-relabel-config'
  scrape_interval: 10s
  static_configs:
  - targets: ['0.0.0.0:8888'] #self obs for collector/prometheus
  metric_relabel_configs:
  - source_labels: [__address__]
    replacement: job_replacement
    target_label: job
...
```

## The collector outputs:
```
2021-12-21T13:12:14.492-0800	debug	scrape/scrape.go:1460	Unexpected error	{"kind": "receiver", "name": "prometheus", "scrape_pool": "metric-relabel-config", "target": "http://0.0.0.0:8888/metrics", "series": "otelcol_exporter_enqueue_failed_log_records{exporter=\"logging\",service_instance_id=\"ec4e2a8b-8b41-458a-9d36-f69e5d86bc33\",service_version=\"latest\"}", "error": "unable to find a target group with job=job_replacement"}
2021-12-21T13:12:14.492-0800	debug	scrape/scrape.go:1248	Append failed	{"kind": "receiver", "name": "prometheus", "scrape_pool": "metric-relabel-config", "target": "http://0.0.0.0:8888/metrics", "error": "unable to find a target group with job=job_replacement"}
2021-12-21T13:12:14.492-0800	warn	internal/metricsbuilder.go:125	Failed to scrape Prometheus endpoint	{"kind": "receiver", "name": "prometheus", "scrape_timestamp": 1640121134490, "target_labels": "map[__name__:up instance:0.0.0.0:8888 job:metric-relabel-config]"}
```

## The prometheus server shows

From (/api/v1/targets/metadata) it the metric with the old metadata:

```json
{"target":{"instance":"0.0.0.0:9090","job":"metric-relabel-config"},"metric":"go_memstats_lookups_total","type":"counter","help":"Total number of pointer lookups.","unit":""}
```

But in the query UI, I can see: 

```
go_memstats_lookups_total{instance="0.0.0.0:9090", job="job_replacement"}
```


# metric relabel configs + instance

## With prometheus config:

```yaml
scrape_configs:
- job_name: 'metric-relabel-config'
  scrape_interval: 10s
  static_configs:
  - targets: ['0.0.0.0:8888'] #self obs for collector/prometheus
  metric_relabel_configs:
  - source_labels: [__address__]
    replacement: instance_replacement
    target_label: instance
...
```

## The collector outputs:
```
2021-12-21T14:09:04.494-0800	debug	scrape/scrape.go:1460	Unexpected error	{"kind": "receiver", "name": "prometheus", "scrape_pool": "metric-relabel-config", "target": "http://0.0.0.0:8888/metrics", "series": "otelcol_exporter_enqueue_failed_log_records{exporter=\"logging\",service_instance_id=\"cc81fd05-120e-4772-8ea1-21fa903bee15\",service_version=\"latest\"}", "error": "unable to find a target with job=metric-relabel-config, and instance=instance_replacement"}
2021-12-21T14:09:04.495-0800	debug	scrape/scrape.go:1248	Append failed	{"kind": "receiver", "name": "prometheus", "scrape_pool": "metric-relabel-config", "target": "http://0.0.0.0:8888/metrics", "error": "unable to find a target with job=metric-relabel-config, and instance=instance_replacement"}
2021-12-21T14:09:04.495-0800	warn	internal/metricsbuilder.go:125	Failed to scrape Prometheus endpoint	{"kind": "receiver", "name": "prometheus", "scrape_timestamp": 1640124544492, "target_labels": "map[__name__:up instance:0.0.0.0:8888 job:metric-relabel-config]"}
```

## The prometheus server shows

From (/api/v1/targets/metadata) it the metric with the old metadata:

```json
{"target":{"instance":"0.0.0.0:9090","job":"metric-relabel-config"},"metric":"go_memstats_lookups_total","type":"counter","help":"Total number of pointer lookups.","unit":""}
```

I didn't see the new metadata.

But in the query UI, I can see: 

```
go_memstats_lookups_total{instance="instance_replacement", job="metric-relabel-config"}
```

So if you tried to look for metric metadata with prometheus + grafana, the metric would be treated as Unknown.


# With relabel_configs + job:

```yaml
scrape_configs:
- job_name: 'relabel-config'
  scrape_interval: 10s
  static_configs:
  - targets: ['0.0.0.0:8888'] #self obs for collector/prometheus
  relabel_configs:
  - source_labels: [__address__]
    replacement: job_replacement
    target_label: job
...
```

## The collector outputs:

```
2021-12-21T14:06:04.344-0800	debug	scrape/scrape.go:1460	Unexpected error	{"kind": "receiver", "name": "prometheus", "scrape_pool": "relabel-config", "target": "http://0.0.0.0:8888/metrics", "series": "otelcol_exporter_enqueue_failed_log_records{exporter=\"logging\",service_instance_id=\"3844962f-c8e6-4bb3-93a0-f9e74530bcb2\",service_version=\"latest\"}", "error": "unable to find a target group with job=job_replacement"}
2021-12-21T14:06:04.344-0800	debug	scrape/scrape.go:1248	Append failed	{"kind": "receiver", "name": "prometheus", "scrape_pool": "relabel-config", "target": "http://0.0.0.0:8888/metrics", "error": "unable to find a target group with job=job_replacement"}
2021-12-21T14:06:04.344-0800	warn	scrape/scrape.go:1203	Appending scrape report failed	{"kind": "receiver", "name": "prometheus", "scrape_pool": "relabel-config", "target": "http://0.0.0.0:8888/metrics", "error": "unable to find a target group with job=job_replacement"}
```

## The prometheus server shows

From (/api/v1/targets/metadata) it the metric with the old metadata:

```json
{"target":{"instance":"0.0.0.0:9090","job":"relabel-config"},"metric":"go_memstats_lookups_total","type":"counter","help":"Total number of pointer lookups.","unit":""}
```

But in the query UI, I can see: 

```
go_memstats_lookups_total{instance="0.0.0.0:9090", job="job_replacement"}
```

# TL;DR

The collector's behavior is always to drop points for which we are unable to find target metadata.  The prometheus server's behavior is to treat them the same as unknown-typed metrics without metadata.  The fix should be similar to https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/5001#issuecomment-925040188.

I'm not sure why our results differed looking at metadata, though.  What endpoint were you querying?

**Comment by gouthamve** [comment]
@dashpole I believe this can be closed now.
